### PR TITLE
Store reported asset events with sensor ticks

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Persists asset materialization/observation/check events emitted from `SensorResult` with the associated instigator tick, giving us the option of displaying and/or linking to the events in the UI - right now they're not persisted at all and not visualized in the UI.

## Test Plan

Unit test ensuring data is saved properly to tick storage.

### CODE REVIEW

**Comment by benpankow** [comment]
* **#23144** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23144?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @benpankow and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_daemon/sensor.py [hunk 1]
```diff
@@ -125,6 +126,10 @@ def update_state(self, status: TickStatus, **kwargs: object):
         skip_reason = cast(Optional[str], kwargs.get("skip_reason"))
         cursor = cast(Optional[str], kwargs.get("cursor"))
         origin_run_id = cast(Optional[str], kwargs.get("origin_run_id"))
+        asset_events = cast(
+            Sequence[EventLogEntry],
+            kwargs.get("asset_events"),
```
Comment (sryza) [inline python_modules/dagster/dagster/_daemon/sensor.py:131]: nit: can be on fewer lines?

File: python_modules/dagster/dagster/_core/scheduler/instigation.py [hunk 2]
```diff
@@ -526,6 +533,10 @@ class TickData(
             ("run_requests", Optional[Sequence[RunRequest]]),  # run requests created by the tick
             ("auto_materialize_evaluation_id", Optional[int]),
             ("reserved_run_ids", Optional[Sequence[str]]),
+            (
+                "asset_events",
+                Optional[Sequence[EventLogEntry]],
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/scheduler/instigation.py:538]: Why store the full event log entry vs. just - say - the `AssetMaterialization` / `AssetObservation` object?

Comment (benpankow) [reply python_modules/dagster/dagster/_core/scheduler/instigation.py:538]: Open to doing so, just means we don't have a way to link to the particular asset event in the catalog

Comment (sryza) [reply python_modules/dagster/dagster/_core/scheduler/instigation.py:538]: How would we do that linking if we had the event log entry?

Comment (benpankow) [reply python_modules/dagster/dagster/_core/scheduler/instigation.py:538]: The `EventLogEntry` has a timestamp, which indexes into the events page right now, e.g. ```https://elementl.dagster.cloud/prod/assets/...?view=events&time=1721621489501```

Comment (benpankow) [reply python_modules/dagster/dagster/_core/scheduler/instigation.py:538]: Updated

**Review by dpeng817 — DISMISSED** [review]
noice

wonder if it makes sense to have the reverse mapping too (in the event metadata, the triggering instigator and tick). That way you could display in the materialization history, "where the materialization came from"

**Review by sryza — COMMENTED** [review]
I think this looks reasonable. I also think it's worth @prha and / or @gibsondan taking a look at it, from a storage perspective, before merging. These events can have arbitrarily-sized metadata dictionaries on them, so definitely possible for them to balloon. Storing the same materialization twice makes me a little nervous.

**Review by gibsondan — CHANGES_REQUESTED** [review]
To sandy's point I am confused why we would not put the storage ID here - similar to how we store the run ID but not the actual beefy run object on these ticks

e.g. if you wipe the materialization it would now still be sitting on this tick, seems bad

**Comment by prha** [comment]
Ditto the previous comments.  I think right now, you could put materializations in `SensorResult` that were never logged/stored.  

I guess that's how it works with `AssetObservation`, but that doesn't seem locked in stone (still experimental).

Could we consider some API where we explicitly log asset events off of the context, and they get accumulated and automatically hung on the sensor result?  Something like:

```
def my_sensor(context, ...):
    context.log_materialization(AssetMaterialization(...))
    context.log_observation(AssetObservation(...))
    return SensorResult(...) # automatically has the asset events logged within the context
```

**Pull request closed without merging**